clang/lib/Driver/ToolChains/Flang.cpp

```diff
@@ -333,6 +333,9 @@ void Flang::AddAMDGPUTargetArgs(const ArgList &Args,
     StringRef Val = A->getValue();
     CmdArgs.push_back(Args.MakeArgString("-mcode-object-version=" + Val));
   }
+
+  const ToolChain &TC = getToolChain();
+  TC.addClangTargetOptions(Args, CmdArgs, Action::OffloadKind::OFK_OpenMP);
```

Comment:
Should there be some kind of warning if these flags are used not with AMDGPU? I don't have a strong opinion here as it is only an fc1 flag.

Reply:
+1 to having better diagnostics

Reply:
Hi,
thanks for the feedback. I would like to share my observations with you:

1. Clang does not verify how we use these flags and it accepts them for non-GPU target.
2. These flags can be reused by other vendors. For example clang adds `mlink-builtin-bitcode` option for OpenMP Nvidia GPU [as well](https://github.com/llvm/llvm-project/blob/ba60d8a11af2cdd7e80e2fd968cdf52adcabf5a1/clang/test/Driver/openmp-offload-gpu.c#L92) .

Reply:
Does that mean that this change would also lead to adding these flags when building for Nvidia GPU with flang?

Reply:
No. My change does not imply any changes for Nvidia GPUs support.

Flang and Clang share the same LLVM backend which consumes generated LLVM IR. For AMD GPU we need to embed bitcode definitions of GPU math functions. AMD toolchain adds all required options to the compiler invocation for AMD GPU and IMO can be reused between Flang and Clang.

I don't know if Nvidia also want to reuse their toolchain between Clang and Flang to fully support OpenMP offloading.

Reply:
> Clang does not verify how we use these flags and it accepts them for non-GPU target.

It's OK to make Flang "stricter" if we believe that's the right thing to do ;-) (I think that generating useful error/warning messages like "don't mix these flags - that's not supporter" would be a good thing)

> IMO can be reused between Flang and Clang

Are there any plans to extract that logic and share it somewhere?

> I don't know if Nvidia also want to reuse their toolchain between Clang and Flang to fully support OpenMP offloading.

Who could be the right person to ask?

Reply:
> It's OK to make Flang "stricter" if we believe that's the right thing to do ;-) (I think that generating useful error/warning messages like "don't mix these flags - that's not supporter" would be a good thing)

Shall I extend https://github.com/llvm/llvm-project/pull/94763 ? I don't use `-fcuda-is-device` anymore. Now, I'm only adding `-mlink-builtin-bitcode` flags to `flang-new -fc1` command. The `-mlink-builtin-bitcode` option was introduced by https://github.com/llvm/llvm-project/pull/94763



> > IMO can be reused between Flang and Clang
> 
> Are there any plans to extract that logic and share it somewhere?

Not yet (at least from my side).  I can return to this topic if there is need to support Clang option by Flang for AMD GPU.



> > I don't know if Nvidia also want to reuse their toolchain between Clang and Flang to fully support OpenMP offloading.
> 
> Who could be the right person to ask?

I don't know. Open-source LLVM Flang meetings can be good  place to ask this question.

Reply:
> Shall I extend https://github.com/llvm/llvm-project/pull/94763 ?

Yes, please.

> Who could be the right person to ask?
> > I don't know. Open-source LLVM Flang meetings can be good place to ask this question.

Did you ask? What feedback did you get?

Reply:
I asked question on flang-compiler slack (openmp/openacc channel). If I get no response, I will raise question on Flang technical community call on Monday.